signing_service/src/signing_service.py

```diff
@@ -0,0 +1,139 @@
+import argparse
+import logging.config
+import os
+import signal
+import socket
+from time import sleep
+
+
+logger = logging.getLogger()
+crash_logger = logging.getLogger('crash')
+
+
+class SigningService:
+    """
+    The Signing Service connects to Middleman as a client but then listens for requests coming from Concent via
+    Middleman. The underlying protocol is TCP and data sent over that is expected to conform to the Wire protocol.
+    """
+
+    def __init__(self, host, port, reconnect_delay) -> None:
+        assert isinstance(host, str)
+        assert isinstance(port, int)
+        assert isinstance(reconnect_delay, int)
+        self.host = host
+        self.port = port
+        self.reconnect_delay = reconnect_delay
+        self.current_reconnect_delay = 0
+        self.socket: socket.socket = None  # type: ignore
+        self.is_connection_established = False
+
+        # Handle shutdown signal.
+        signal.signal(signal.SIGTERM, self._terminate_connection_and_exit)  # type: ignore
+
+        self._run()
+
+    def _run(self) -> None:
+        """
+        Handles main connection loop and its exceptions.
+
+        If connection is interrupted due to a failure, it should try to reconnect.
+        If connection is closed by other side, exit gracefully.
+        If a shutdown signal or KeyboardInterrupt is caught, exit gracefully.
+        If there was an unrecognized exception, it logs it and report to Sentry, then reraise and crash.
+        """
+        while True:
+            try:
+                if self.is_connection_established is False:
+                    self._connect()
+            except socket.error as exception:
+                logger.error(f'Socket error occurred: {exception}')
```

Comment:
Do all possible `socket.error` exceptions represent connection errors? I think we want to handle only a few specific `socket.errno` values and reraise all the others (handle them like the `Exception` handler below does).

Reply:
Will this list be ok ?
```
74 Bad message
86 Streams pipe error
87 Too many users
90 Message too long
100 Network is down
101 Network is unreachable
102 Network dropped connection on reset
103 Software caused connection abort
104 Connection reset by peer
108 Cannot send after transport endpoint shutdown
110 Connection timed out
111 Connection refused
112 Host is down
113 No route to host
```

Reply:
Will this list be ok ?
```
74 Bad message
86 Streams pipe error
87 Too many users
90 Message too long
100 Network is down
101 Network is unreachable
102 Network dropped connection on reset
103 Software caused connection abort
104 Connection reset by peer
108 Cannot send after transport endpoint shutdown
110 Connection timed out
111 Connection refused
112 Host is down
113 No route to host
```

Reply:
Seems fine, but I'm not sure about these two:
```
74 Bad message
87 Too many users
```

In general I'd err on the side of catching too little rather than too much. If you catch something you shouldn't, we may miss a bug. If you don't catch something you should, the only difference will be that the service will crash, send a report and get restarted by Kubernetes instead of retrying on its own.

Reply:
I agree, I will create a list without those two.